File: javascript/Phylo/Forest/Node.js
(function(){
function Node (args) {
	if (args==null) args = {};
	args["tag"] = "node";
    this.XMLWritable(args);
    this.children   = [];
    this._type      = Phylo.Util.CONSTANT._NODE_;
    this._container = Phylo.Util.CONSTANT._TREE_;
    return this;
}
Phylo.Forest.Node = Node;
copyPrototypeMI(Phylo.Forest.Node,[Phylo.Util.XMLWritable,Phylo.Taxa.TaxonLinker]);
var proto = Phylo.Forest.Node.prototype;

proto.set_parent = function (parent) {
    this.parent = parent;
    return this;
};

proto.set_child = function (child) {
    this.children.push(child);
    return this;
};

proto.set_first_daughter = function (child) {
    this.children.unshift(child);
    return this;
};

proto.set_last_daughter = function (child) {
    this.children.push(child);
    return this;
};

proto.set_previous_sister = function (sister) {
    var parent = this.get_parent();
    if ( parent != null ) {
        var children = parent.get_children();
        for ( var i = 0; i < children.length; i++ ) {
            if ( children[i].get_id == this.get_id ) {
                if ( i > 0 ) {
                    for ( var j = children.length; j >= i; j-- ) {
                        children[j] = children[j-1];
                    }
                    children[i-1] = sister;
                }
                else {
                    children.unshift(sister);
                }
            }
        }
    }
    return this;
};

proto.set_next_sister = function (sister) {
    var parent = this.get_parent();
    if ( parent != null ) {
        var children = parent.get_children();
        for ( var i = 0; i < children.length; i++ ) {
            if ( children[i].get_id == this.get_id ) {
                for ( var j = children.length; j > i; j-- ) {
                    children[j+1] = children[j];
                }
                children[i+1] = sister;
            }
        }
    }
    return this;
};

proto.set_branch_length = function (branch_length) {
    this.branch_length = Number(branch_length);
    return this;
};

proto.set_node_below = function () {
    var new_parent = new Phylo.Forest.Node();
    var parent = this.get_parent();
    if ( parent != null ) {
        parent.set_child(new_parent);
        new_parent.set_child(this);
        new_parent.set_parent(parent);
        this.set_parent(new_parent);
        parent.prune_child(this);
    }
    else {
        this.set_parent(new_parent);
        new_parent.set_child(this);
    }
    return new_parent;   
};

proto.prune_child = function (child) {
    for ( var i = 0; i < this.children.length; i++ ) {
        if ( this.children[i].get_id() == child.get_id() ) {
            this.children.splice(i,1);
            return this;
        }
    }
    return this;
};

proto.get_parent = function () {
    return this.parent;
};

proto.get_children = function () {
    return this.children;
};

proto.get_first_daughter = function () {
    if ( this.children[0] != null ) {
    	return this.children[0];
    }
    else {
    	return null;
    }
};

proto.get_last_daughter = function () {
    for ( var i = 0; i < this.children.length; i++ ) {
        if ( this.children[i+1] == null ) {
            return this.children[i];
        }
    }
    return null;
};

proto.get_next_sister = function () {
    var parent = this.get_parent();
    if ( parent != null ) {
        var siblings = parent.get_children();
        for ( var i = 0; i < siblings.length; i++ ) {
            if ( siblings[i].get_id() == this.get_id() ) {
                if ( i+1 <= siblings.length ) {
                	if ( siblings[i+1] != null ) {
                    	return siblings[i+1];
                	}
                	else {
                		return null;
                	}
                }
            }
        }
    }
    return null;
};

proto.get_previous_sister = function () {
    var parent = this.get_parent();
    if ( parent != null ) {
        var siblings = parent.get_children();
        for ( var i = 0; i < siblings.length; i++ ) {
            if ( siblings[i].get_id() == this.get_id() ) {
                if ( i > 0 ) {
                	if ( siblings[i-1] != null ) {
                    	return siblings[i-1];
                	}
                	else {
                		return null;
                	}
                }
            }
        }
    }
    return null;
};

proto.get_branch_length = function () {
    return this.branch_length;
};

proto.get_descendants = function () {
	var descendants = new Array();
	this.visit_depth_first( {
		"pre" : function(node) {
			descendants.push(node);
		}
	} );
	descendants.shift(); // remove starting node
	return descendants;
};

proto.get_ancestors = function () {
	var ancestors = new Array();
	var node = this.get_parent();
	while ( node != null ) {
		ancestors.push(node);
		node = node.get_parent();
	}
	return ancestors;
};

proto.is_terminal = function () {
    return this.children.length == 0 ? true : false;
};

proto.is_internal = function () {
    return this.children.length == 0 ? false : true;
};

proto.is_root = function () {  
    return this.get_parent() == null ? true : false;
};

proto.is_first = function () { 
    return this.get_previous_sister() == null ? true : false;
};

proto.is_last = function () { 
    return this.get_next_sister() == null ? true : false;
};

proto.is_descendant_of = function (ancestor) {
    var node = this;
    while ( node != null ) {
        node = node.get_parent();
        if ( node.get_id() == ancestor.get_id() ) {
            return true;
        }
    }
    return false;
};

proto.is_ancestor_of = function (descendant) {
    while( descendant != null ) {
        descendant = descendant.get_parent();
        if ( descendant.get_id() == this.get_id() ) {
            return true;
        }
    }
    return false;
};

proto.calc_nodes_to_root = function () {
	var n = 0;
	var node = this;
	while(node) {
		node = node.get_parent();
		n++;
	}
	return n;
};

proto.calc_max_path_to_tips = function () {
	var maxpath = 0;
	var root = this;
	this.visit_depth_first({
		'pre' : function (n) {
			if ( n.is_terminal() ) {
				var node = n;
				var path = 0;
				while( node.get_id() != root.get_id() ) {
					path += node.get_branch_length();
					node = node.get_parent();
				}
				if ( path > maxpath ) maxpath = path;
			}
		}
	});
	return maxpath;
};

proto.calc_max_nodes_to_tips = function () {
	var maxnodes = 0;
	var root = this;
	this.visit_depth_first({
		'pre' : function (n) {
			if ( n.is_terminal() ) {
				var node = n;
				var nodes = 0;
				while( node.get_id() != root.get_id() ) {
					nodes++;
					node = node.get_parent();
				}
				if ( nodes > maxnodes ) maxnodes = nodes;
			}
		}
	});
	return maxnodes;
};

proto.visit_depth_first = function (args) {
    if (args==null) args = {};
    if (args["pre"]!=null) args["pre"](this);
    
    // daughter related code refs
    var fd = this.get_first_daughter();
    if ( fd != null ) {
        if (args["pre_daughter"]!=null) args["pre_daughter"](this);
        fd.visit_depth_first(args);
        if (args["post_daughter"]!=null) args["post_daughter"](this);
    }
    else {
        if (args["no_daughter"]!=null) args["no_daughter"](this);
    }
    
    // in-order code ref
    if (args["in"]!=null) args["in"](this);
    
    // sister related code refs
    var ns = this.get_next_sister();
    if ( ns != null ) {
        if (args["pre_sister"]!=null) args["pre_sister"](this);
        ns.visit_depth_first(args);
        if (args["post_sister"]!=null) args["post_sister"](this);
    }
    else {
        if (args["no_sister"]!=null) args["no_sister"](this);
    }    
        
    if (args["post"]!=null) args["post"](this);
};

var newick_string = new String();
proto.to_newick = function (args) {
    var name = this.get_name();
    var branch_length = this.get_branch_length();
    if ( this.is_internal() ) {
        newick_string += '(';
        this.get_first_daughter().to_newick(args);
        newick_string += ')';
    }
    if ( this.get_name() != null ) {
    	newick_string += this.get_name();
    }
    else if ( this.get_taxon() != null && this.get_taxon().get_name() != null ) {
    	newick_string += this.get_taxon().get_name();
    }
    if ( this.get_branch_length() != null ) {
    	newick_string += ':' + this.get_branch_length();
    }
    if ( args != null && args['nhxkeys'] != null ) {
    	var nhx = '[&&NHX:';
    	for ( var i = 0; i < args['nhxkeys'].length; i++ ) {
    		nhx += args['nhxkeys'][i] + '=';
    		nhx += this.get_generic(args['nhxkeys'][i]) + ':';
    	}
    	nhx += ']';
    	newick_string += nhx;
    }
    if ( this.is_root() ) {
        var result = newick_string + ';';
        newick_string = new String();
        return result;
    }
    else if ( this.get_next_sister() != null ) {
        newick_string += ',';
        this.get_next_sister().to_newick(args);
    }
    return null;
};

proto.to_xml = function () {
	var nodes = new Array();
	var desc = this.get_descendants();
	nodes.push(this);
	for ( var i = 0; i < desc.length; i++ ) {
		nodes.push(desc[i]);
	}
	var xml = '';
	
	//first write out the node elements
	for ( var i = 0; i < nodes.length; i++ ) {
		var taxon = nodes[i].get_taxon();
		if ( taxon != null ) {
			nodes[i].set_attributes( { 'otu' : taxon.get_xml_id() } );
		}
		if ( nodes[i].is_root() ) {
			nodes[i].set_attributes( { 'root' : true } );
		}
		xml += nodes[i].get_xml_tag(true);
	}
	
	//then the rootedge?
	var root_length = nodes.shift().get_branch_length();
	if ( root_length != null && root_length != 0 ) {
		var target = this.get_xml_id();
		var id = "edge" + this.get_id();
		xml += '<rootedge target="' + target + '" id="' + id + '" length="' + root_length + '"/>';
	}
	
	//then the subtended edges
	for ( var i = 0; i < nodes.length; i++ ) {
		var source = nodes[i].get_parent().get_xml_id();
		var target = nodes[i].get_xml_id();
		var id = "edge" + nodes[i].get_id();
		var branch_length = nodes[i].get_branch_length();
		xml += '<edge source="' + source;
		xml += '" target="' + target;
		xml += '" id="' + id;
		if ( branch_length != null ) xml += '" length="' + branch_length;
		xml += '"/>';
	}		
		
	return xml;
}
})()

File: nexml/java/src/org/biophylo/Forest/Node.java
package org.biophylo.Forest;
import org.biophylo.Util.*;
import org.biophylo.Containable;
import org.biophylo.Visitor;
import org.biophylo.Taxa.*;
import org.biophylo.Mediators.*;
import org.biophylo.Util.Exceptions.*;
import org.w3c.dom.*;

import java.util.*;
import java.math.*;

public class Node extends Containable implements TaxonLinker {
	private static Logger logger = Logger.getInstance();
	protected Node parent;
	Vector children;
	protected double branch_length; 
	private static TaxaMediator taxaMediator = TaxaMediator.getInstance();
	
	/**
	 * 
	 */
	public Node () {
		super();
		this.type = CONSTANT.NODE;
		this.container = CONSTANT.TREE;
		this.children = new Vector();
		this.tag = "node";
		this.parent = null;
	}
	
	/* (non-Javadoc)
	 * @see org.biophylo.Taxa.TaxonLinker#unsetTaxon()
	 */
	public void unsetTaxon() {
		int linkerId = this.getId();
		taxaMediator.removeLink(-1, linkerId);
	}
	
	/* (non-Javadoc)
	 * @see org.biophylo.Taxa.TaxonLinker#setTaxon(org.biophylo.Taxa.Taxon)
	 */
	public void setTaxon (Taxon taxon) {
		int taxonId = taxon.getId();
		int linkerId = this.getId();
		taxaMediator.setLink(taxonId, linkerId);
	}
	
	/* (non-Javadoc)
	 * @see org.biophylo.Taxa.TaxonLinker#getTaxon()
	 */
	public Taxon getTaxon () {
		int linkerId = this.getId();
		return (Taxon)taxaMediator.getLink(linkerId);
	}
	
	/**
	 * @param parent
	 */
	public void setParent(Node parent) { // XXX node mediator
		this.parent = parent;
	}
	
	/**
	 * @return
	 */
	public Node getParent() { // XXX node mediator
		return this.parent;
	}	
	
	/**
	 * @param child
	 */
	public void setChild(Node child) { // XXX node mediator
		this.children.add(child);
	}	
	
	/**
	 * @return
	 */
	public Node[] getChildren() { // XXX node mediator
		Node[] result = new Node[this.children.size()];
		this.children.copyInto(result);
		return result;
	}
	
	/**
	 * @param index
	 * @return
	 */
	public Node getChild(int index) {
		return this.getChildren()[index];
	}
	
	/**
	 * @return
	 */
	public Node getFirstDaughter() {
		Node[] children = this.getChildren();
		if ( children.length != 0 ) {
			return children[0];
		}
		else {
			return null;
		}
	}
	
	/**
	 * @return
	 */
	public Node getLastDaughter() {
		Node[] children = this.getChildren();
		if ( children.length != 0 ) {
			return children[children.length-1];
		}
		else {
			return null;
		}
	}
	
	/**
	 * @return
	 */
	public Node getNextSister () {
		if ( ! this.isRoot() ) {
			Node[] sisters = this.getSisters();
			int numSibs = sisters.length;
			for ( int i = 0; i < numSibs; i++ ) {
				if ( sisters[i].getId() == this.getId() && i < numSibs - 1) {
					int nsIndex = i+1;
					return sisters[nsIndex];
				}
			}
		}
		return null;
	}
	
	/**
	 * @return
	 */
	public Node getPreviousSister() {
		if ( ! this.isRoot() ) {
			Node[] siblings = this.getParent().getChildren();
			for ( int i = 0; i < siblings.length; i++ ) {
				if ( siblings[i].getId() == this.getId() && i > 0 ) {
					return siblings[i-1];
				}
			}
		}
		return null;		
	}
	
	/**
	 * @return
	 */
	public Node[] getAncestors () {
		Vector ancestors = new Vector();
		Node n = this.getParent();
		while ( n != null ) {
			ancestors.add(n);
			n = n.getParent();
		}
		Node[] result = new Node[ancestors.size()];
		ancestors.copyInto(result);
		return result;
	}
	
	/**
	 * @return
	 */
	public Node[] getSisters(){
		Node parent = this.getParent();
		if ( parent != null ) {
			return parent.getChildren();
		}
		else {
			return null;
		}		
	}
	
	/**
	 * @return
	 */
	public Node[] getDescendants() {
		class NodeCollector implements Visitor {
			public Vector result = new Vector();
			public void visit (Containable node) {
				result.add(node);
			}
		}
		NodeCollector nc = new NodeCollector();
		HashMap args = new HashMap();
		args.put("pre", nc);
		this.visitDepthFirst(args);
		nc.result.remove(0);
		Node[] result = new Node[nc.result.size()];
		nc.result.copyInto(result);
		return result;
	}
	
	/**
	 * @return
	 */
	public Node[] getTerminals() {		
		Vector terminals = new Vector();
		Node[] desc = this.getDescendants();
		for ( int i = 0; i < desc.length; i++ ) {
			if ( desc[i].isTerminal() ) {
				terminals.add(desc[i]);
			}
		}
		Node[] result = new Node[terminals.size()];
		terminals.copyInto(result);
		return result;
	}
	
	/**
	 * @return
	 */
	public Node[] getInternals() {
		Vector internals = new Vector();
		Node[] desc = this.getDescendants();
		for ( int i = 0; i < desc.length; i++ ) {
			if ( desc[i].isInternal() ) {
				internals.add(desc[i]);
			}
		}
		Node[] result = new Node[internals.size()];
		internals.copyInto(result);
		return result;
	}	
	
	/**
	 * @param node
	 * @return
	 */
	public Node getMrca(Node node) {
		Node[] myAnc = this.getAncestors();
		Node[] otherAnc = this.getAncestors();
		Node mrca = null;
		MRCA: for ( int i = 0; i < myAnc.length; i++) {
			for ( int j = 0; j < otherAnc.length; j++ ) {
				if ( myAnc[i].getId() == otherAnc[j].getId() ) {
					mrca = myAnc[i];
					break MRCA;
				}
			}
		}
		return mrca;
	}
	
	/**
	 * @return
	 */
	public Node getLeftmostTerminal () {
		Node node = this;
		Node lmt = null;
		while (node != null ) {
			lmt = node;
			node = node.getFirstDaughter();
		}
		return lmt;
	}
	
	/**
	 * @return
	 */
	public Node getRightmostTerminal() {
		Node node = this;
		Node rmt = null;
		while (node != null ) {
			rmt = node;
			node = node.getLastDaughter();
		}
		return rmt;		
	}
	
	/**
	 * @param branch_length
	 */
	public void setBranchLength(double branch_length) {
		this.branch_length = branch_length;
	}
	
	/**
	 * @return
	 */
	public double getBranchLength() {
		return this.branch_length;
	}
	
	/**
	 * @return
	 */
	public boolean isTerminal() {
		return this.getChildren().length == 0 ? true : false;
	}
	
	/**
	 * @return
	 */
	public boolean isInternal () {
		return ! this.isTerminal();
	}

	/**
	 * @return
	 */
	public boolean isRoot() {
		if ( this.getParent() == null ) {
			return true;
		}
		else {
			return false;
		}
	}
	
	/**
	 * @return
	 */
	public boolean isFirst() {
		return this.getPreviousSister() == null ? true : false;
	}
	
	/**
	 * @return
	 */
	public boolean isLast() {
		return this.getNextSister() == null ? true : false;
	}
	
	/**
	 * @param ancestor
	 * @return
	 */
	public boolean isDescendantOf(Node ancestor) {
		Node[] anc = this.getAncestors();
		for ( int i = 0; i < anc.length; i++ ) {
			if ( anc[i].getId() == ancestor.getId() ) {
				return true;
			}
		}
		return false;
	}
	
	/**
	 * @param descendant
	 * @return
	 */
	public boolean isAncestorOf(Node descendant) {
		return descendant.isDescendantOf(this);
	}
	
	/**
	 * @param sister
	 * @return
	 */
	public boolean isSisterOf(Node sister) {
		Node[] sisters = this.getSisters();
		for ( int i = 0; i < sisters.length; i++ ) {
			if ( sisters[i].getId() == sister.getId() && sisters[i].getId() != this.getId() ) {
				return true;
			}
 		}
		return false;
	}
	
	/**
	 * @param parent
	 * @return
	 */
	public boolean isChildOf(Node parent) {
		if ( ! this.isRoot() ) {
			return this.getParent().getId() == parent.getId();
		}
		return false;
	}
	
	/**
	 * @param child
	 * @return
	 */
	public boolean isParentOf(Node child) {
		return child.isChildOf(this);
	}
	
	/**
	 * @param nodes
	 * @return
	 */
	public boolean isOutgroupOf(Node[] nodes) {
		for ( int i = 0; i < nodes.length; i++ ) {
			for ( int j = i+1; j < nodes.length; j++ ) {
				Node mrca = nodes[i].getMrca(nodes[j]);
				if ( mrca.isAncestorOf(this) ) {
					return true;
				}
			}
		}
		return true;	
	}
	
	/**
	 * @return
	 */
	public double calcPathToRoot() {
		double path = 0;
		Node[] anc = this.getAncestors();
		for ( int i = 0; i < anc.length; i++ ) {
			path += anc[i].getBranchLength();
		}
		return path;
	}
	
	/**
	 * @return
	 */
	public int calcNodesToRoot() {
		Node[] anc = this.getAncestors();
		if ( anc != null ) {
			return anc.length;
		}
		else {
			return 0;
		}
	}
	
	/**
	 * @return
	 */
	public int calcMaxNodesToTips() {
		Node[] tips = this.getTerminals();
		int maxnodes = 0;
		for ( int i = 0; i < tips.length; i++ ) {
			Node focal = tips[i];
			int dist = 0;
			while ( focal.getId() != this.getId() ) {
				dist++;
				focal = focal.getParent();
			}
			if ( dist > maxnodes) {
				maxnodes = dist;
			}
		}
		return maxnodes;
	}
	
	/**
	 * @return
	 */
	public int calcMinNodesToTips() {
		Node[] tips = this.getTerminals();
		int minnodes = 0;
		boolean uninit = true;
		for ( int i = 0; i < tips.length; i++ ) {
			Node focal = tips[i];
			int dist = 0;
			while ( focal.getId() != this.getId() ) {
				dist++;
				focal = focal.getParent();
			}
			if ( dist < minnodes || uninit ) {
				minnodes = dist;
				uninit = true;
			}
		}
		return minnodes;
	}
	
	/**
	 * @return
	 */
	public double calcMaxPathToTips() {
		double maxpath = 0;
		Node[] desc = this.getDescendants();
		for ( int i = 0; i < desc.length; i++ ) {
			Node focal = desc[i];
			double dist = 0;
			while ( focal.getId() != this.getId() ) {
				dist += focal.getBranchLength();
				focal = focal.getParent();
			}
			if ( dist > maxpath ) {
				maxpath = dist;
			}
		}
		return maxpath;
	}
	
	/**
	 * @return
	 */
	public double calcMinPathToTips () {
		double minpath = 0;
		boolean uninit = true;
		Node[] desc = this.getDescendants();
		for ( int i = 0; i < desc.length; i++ ) {
			Node focal = desc[i];
			double dist = 0;
			while ( focal.getId() != this.getId() ) {
				dist += focal.getBranchLength();
				focal = focal.getParent();
			}
			if ( dist < minpath || uninit ) {
				minpath = dist;
				uninit = true;
			}
		}
		return minpath;
	}
	
	/**
	 * @param node
	 * @return
	 */
	public double calcPatristicDistance (Node node) {
		double dist = 0;
		Node mrca = this.getMrca(node);
		int mrcaId = mrca.getId();
		Node me = this;
		while ( me.getId() != mrcaId ) {
			dist += me.getBranchLength();
			me = me.getParent();
		}
		while ( node.getId() != mrcaId ) {
			dist += node.getBranchLength();
			node = node.getParent();
		}
		return dist;
	}
	
	/**
	 * @param node
	 * @return
	 */
	public int calcNodalDistance (Node node) {
		int dist = 0;
		Node mrca = this.getMrca(node);
		int mrcaId = mrca.getId();
		Node me = this;
		while ( me.getId() != mrcaId ) {
			dist++;
			me = me.getParent();
		}
		while ( node.getId() != mrcaId ) {
			dist++;
			node = node.getParent();
		}
		return dist;
	}
	
	/**
	 * @param args
	 */
	public void visitDepthFirst(HashMap args) {
		this.process(args,"pre");
		Node fd = this.getFirstDaughter();
		if ( fd != null ) {
			this.process(args,"pre_daughter");
			fd.visitDepthFirst(args);
			this.process(args,"post_daughter");
		}
		else {
			this.process(args,"no_daughter");
		}
		this.process(args, "in");
		Node ns = this.getNextSister();
		if ( ns != null ) {
			this.process(args, "pre_sister");
			ns.visitDepthFirst(args);
			this.process(args, "post_sister");
		}
		else {
			this.process(args, "no_sister");
		}
		this.process(args, "post");
	}	
	
	/**
	 * @return
	 */
	public String toNewick() {
		final StringBuffer newick = new StringBuffer();
		class PreDaughterWriter implements Visitor {
			public void visit (Containable node) {
				newick.append("(");
			}
		}
		class PreSisterWriter implements Visitor {
			public void visit (Containable node) {
				newick.append(",");
			}			
		}
		class PostDaughterWriter implements Visitor {
			public void visit (Containable node) {
				newick.append(")");
			}			
		}
		class InWriter implements Visitor {
			public void visit (Containable node) {
				String name = node.getName();
				double bl = ((Node)node).getBranchLength();
				if ( name != null ) {
					newick.append(name);
				}
				newick.append(":");
				newick.append(bl+"");
			}			
		}
		HashMap args = new HashMap();
		args.put("pre_daughter", new PreDaughterWriter());
		args.put("pre_sister", new PreSisterWriter());
		args.put("post_daughter", new PostDaughterWriter());
		args.put("in", new InWriter());
		this.visitDepthFirst(args);
		newick.append(";");
		return newick.toString();
	}
	
	/**
	 * @param roundAsInt
	 * @return
	 */
	public Element edgeToXmlElement (boolean roundAsInt) {
		HashMap attrs = new HashMap();
		attrs.put("id", "edge" + getId());
		attrs.put("target", getXmlId());
		String tagName = null;
		if ( isRoot() ) {
			if ( getBranchLength() != 0 ) {
				tagName = "rootedge";
			}
			else {
				return null;
			}
		}
		else {
			tagName = "edge";
			attrs.put("source", getParent().getXmlId() );
		}
		if ( roundAsInt ) {
			attrs.put("length", ""+Math.round(getBranchLength()));
		}
		else {
			attrs.put("length", ""+getBranchLength());
		}
		return createElement(tagName,attrs,getDocument());
	}	
	
	/**
	 * @param args
	 * @param key
	 */
	private void process(HashMap args,String key) {
		if ( args.containsKey(key) ) {
			if ( args.get(key) instanceof Visitor ) {
				((Visitor)args.get(key)).visit(this);
			}
		}
	}
	
	/* (non-Javadoc)
	 * @see org.biophylo.Base#finalize()
	 */
	protected void finalize() throws Throwable {
	  //do finalization here
		taxaMediator.removeLink(-1, this.getId());
	  super.finalize(); //not necessary if extending Object.
	} 	

}
